src/qibo/models/variational.py

```diff
@@ -659,3 +663,100 @@ def _loss(params, falqon, hamiltonian):
+    """
+
+    def __init__(self, nqubits, accelerators=None, density_matrix=False):
+        super().__init__(nqubits, accelerators, density_matrix)
+
+    def _get_initparams(self):
+        """Retrieve parameter values or objects directly from gates"""
+
+        params = []
+        for gate in self.queue:
+            if isinstance(gate, (gates.ParametrizedGate)):
+                try:
+                    params.append(gate.initparams)
+                except AttributeError:
+                    params.append(gate.parameters)
+
+        if isinstance(params[0], (float, int, tuple)):
+            params = self.get_parameters()
+            if isinstance(params[0], tuple):
+                params = np.array([val for t in params for val in t])
+
+        return params
+
+    def _get_train_params(self):
+        # for array
+        if isinstance(self.initparams, np.ndarray):
+            return self.initparams
+
+        # for Parameter objects
+        else:
+            params = []
+            for param_object in self.initparams:
+                # update trainable params and retrieve gate param
+                params.extend(param_object._trainable)
+
+            return params
+
+    def set_variational_parameters(self, input_params, feature=None):
+        """Retrieve gate parameters based on initial parameter values given to gates
+        Args:
+            feature (int or list): input feature if embedded in Parameter lambda function
+
+        Returns:
+            (list or np.ndarray) gate parameters
+        """
+
+        # for array
+        if isinstance(self.initparams, np.ndarray):
+            gate_params = self.initparams
+
+        # for Parameter objects
+        else:
+            gate_params = []
+            count = 0
+            for param_object in self.initparams:
+                trainable = input_params[count : count + param_object.nparams]
+                count += param_object.nparams
+                # update trainable params and retrieve gate param
+                print(trainable, feature)
+                param_object.update_parameters(trainable, feature)
+                gate_params.append(param_object())
+
+        self.set_parameters(gate_params)
+
+    def add(self, gate):
+        """Add a gate to a given queue.
+
+        Args:
+            gate (:class:`qibo.gates.Gate`): the gate object to add.
+                See :ref:`Gates` for a list of available gates.
+                `gate` can also be an iterable or generator of gates.
+                In this case all gates in the iterable will be added in the
+                circuit.
+
+        Returns:
+            If the circuit contains measurement gates with ``collapse=True``
+            a ``sympy.Symbol`` that parametrizes the corresponding outcome.
+        """
+        res = super().add(gate)
+        if isinstance(gate, gates.ParametrizedGate):
+            self.initparams = self._get_initparams()
```

Comment:
This is one of the things I don't fully get. This is essentially an update of `self.initparams`, but you're overwriting it every time you `add()` a gate.

Can't you just compute `self.initparams` once before execution?
Is there a time in which you need `self.initparams` before `add`ing one more gate?

Reply:
Maybe you know of a solution to this. Basically, I need to get the initparams once the circuit is fully built. However, there is no way of knowing when that is the case, the user could keep adding gates. Since I can't make assumptions about when the circuit is finished being built, I need to call self._get_initparams after each gate is added to the circuit